cin_validator/rules/cin2022_23/rule_8940.py

```diff
@@ -0,0 +1,262 @@
+
+    #  Create dataframes which only have rows with CP plans, and which should have one plan per row.
+    df_cpp = df_cpp[df_cpp[CPPstartDate].notna()]
+    df_cpp2 = df_cpp2[df_cpp2[CPPstartDate].notna()]
+    print("\n df_cpp\n", df_cpp)
+    #  Merge tables to test for overlaps
+    df_merged = df_cpp.merge(
+        df_cpp2,
+        on=["LAchildID"],
+        how="left",
+        suffixes=("_cpp", "_cpp2"),
+    )
+
+    # Exclude rows where the CPPID is the same on both sides
+    df_merged = df_merged[(df_merged["CPPID_cpp"] != df_merged["CPPID_cpp2"])]
+    print("\n df_merged exclude\n", df_merged)
+    # Determine whether CPP overlaps another CPP
+    cpp_started_after_start = (
+        df_merged["CPPstartDate_cpp"] >= df_merged["CPPstartDate_cpp2"]
+    )
+    cpp_started_before_end = (
+        df_merged["CPPstartDate_cpp"] < df_merged["CPPendDate_cpp2"]
+    ) & df_merged["CPPendDate_cpp2"].notna()
+    cpp_started_before_refdate = (
+        df_merged["CPPstartDate_cpp"] < reference_date
+    ) & df_merged["CPPendDate_cpp2"].isna()
+
+    df_merged = df_merged[
+        cpp_started_after_start & (cpp_started_before_end | cpp_started_before_refdate)
+    ].reset_index()
+    print("\n df_merged filtered\n", df_merged)
+    # create an identifier for each error instance.
+    # In this case, the rule is checked for each CPPstartDate, in each CPplanDates group (differentiated by CP dates), in each child (differentiated by LAchildID)
+    # So, a combination of LAchildID, CPPstartDate and CPPreviewDate identifies and error instance.
+    df_merged["ERROR_ID"] = tuple(
+        zip(df_merged[LAchildID], df_merged["CPPID_cpp"], df_merged["CPPID_cpp2"])
+    )
+    print("\n df_merged zipped\n", df_merged)
+    # The merges were done on copies of cpp_df and reviews_df so that the column names in dataframes themselves aren't affected by the suffixes.
+    # we can now map the suffixes columns to their corresponding source tables such that the failing ROW_IDs and ERROR_IDs exist per table.
+    df_cpp_issues = (
+        df_cpp.merge(df_merged, left_on="ROW_ID", right_on="ROW_ID_cpp")
+        .groupby("ERROR_ID", group_keys=False)["ROW_ID"]
+        .apply(list)
+        .reset_index()
+    )
+    df_cpp2_issues = (
+        df_cpp2.merge(df_merged, left_on="ROW_ID", right_on="ROW_ID_cpp2")
+        .groupby("ERROR_ID", group_keys=False)["ROW_ID"]
+        .apply(list)
+        .reset_index()
+    )
+    print("\n df_cpp_issues\n", df_cpp_issues)
+    print("\n df_cpp2_issues\n", df_cpp2_issues)
+
+    # Ensure that you maintain the ROW_ID, and ERROR_ID column names which are shown above. They are keywords in this project.
+    rule_context.push_type_2(
+        table=ChildProtectionPlans, columns=[CPPstartDate], row_df=df_cpp_issues
+    )
+    rule_context.push_type_2(
+        table=ChildProtectionPlans,
+        columns=[CPPstartDate, CPPendDate],
+        row_df=df_cpp2_issues,
+    )
```

Comment:
This is one of the scenarios where both push_type2 and push_type3 work. 

Still, if you replace these with push_type3, then the rule will be classified properly as a type3 rule in case we want to filter later one. Could you do that?.

Reply:
Done!